### PR TITLE
[Refactor] limit Local exchange buffer size by memory usage rather than by row count

### PR DESCRIPTION
**Pull request by fzhedu**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

local exchange buffer's size limits the sink's speed, causing low memory usage in some cases. This PR will increase CPU usage from 40+% to 90+% for some agg queries limited by network.

besides, it alleviates the limit of exchange's scalability with dop.
```
select host_name() h,exchange_speed(ss_customer_sk,ss_quantity,ss_sales_price), count(*) from store_sales where ss_customer_sk is not null group by ss_customer_sk;
```
<meta charset="utf-8"><div data-page-id="CDC9dn54Voz89KxfQGScj6zanMb" data-docx-has-block-data="true"><div>

dop | 0 | 16 | 32
-- | -- | -- | --
before | 41.53 | 57.72 | 141.95
after | 44.253504 | 44.71 | 51.35

</div></div><span data-lark-record-data="{&quot;isCut&quot;:false,&quot;rootId&quot;:&quot;CDC9dn54Voz89KxfQGScj6zanMb&quot;,&quot;parentId&quot;:&quot;HOYad0CMuo2UysxQzSUcvXaVnBb&quot;,&quot;blockIds&quot;:[640,642,644,646,648,650,652,654,656,658,660,662],&quot;recordIds&quot;:[&quot;HOYad0CMuo2UysxQzSUcvXaVnBb&quot;],&quot;recordMap&quot;:{&quot;NgO0dCYuSo0SMQxSO9TcDO6gn0e&quot;:{&quot;id&quot;:&quot;NgO0dCYuSo0SMQxSO9TcDO6gn0e&quot;,&quot;snapshot&quot;:{&quot;parent_id&quot;:&quot;HOYad0CMuo2UysxQzSUcvXaVnBb&quot;,&quot;type&quot;:&quot;table_cell&quot;,&quot;children&quot;:[&quot;AWyIdU0aYossGwxbHVXcNszSnbN&quot;],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;}},&quot;AWyIdU0aYossGwxbHVXcNszSnbN&quot;:{&quot;id&quot;:&quot;AWyIdU0aYossGwxbHVXcNszSnbN&quot;,&quot;snapshot&quot;:{&quot;parent_id&quot;:&quot;NgO0dCYuSo0SMQxSO9TcDO6gn0e&quot;,&quot;type&quot;:&quot;text&quot;,&quot;children&quot;:[],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;,&quot;text&quot;:{&quot;initialAttributedTexts&quot;:{&quot;text&quot;:{&quot;0&quot;:&quot;dop&quot;},&quot;attribs&quot;:{&quot;0&quot;:&quot;*0*1+3&quot;}},&quot;apool&quot;:{&quot;numToAttrib&quot;:{&quot;0&quot;:[&quot;abbreviation-data&quot;,&quot;{\&quot;id\&quot;:\&quot;ac4ee74c-f062-498b-8e46-50839c546556\&quot;,\&quot;abbr_ids\&quot;:\&quot;enterprise_7160956291072966660\&quot;,\&quot;is_visible\&quot;:1,\&quot;is_first\&quot;:1}&quot;],&quot;1&quot;:[&quot;author&quot;,&quot;7115230360346050564&quot;]},&quot;nextNum&quot;:2}},&quot;folded&quot;:false}},&quot;UkcMdKC2kouIa8xiSwzccorWnad&quot;:{&quot;id&quot;:&quot;UkcMdKC2kouIa8xiSwzccorWnad&quot;,&quot;snapshot&quot;:{&quot;parent_id&quot;:&quot;HOYad0CMuo2UysxQzSUcvXaVnBb&quot;,&quot;type&quot;:&quot;table_cell&quot;,&quot;children&quot;:[&quot;GmCWdASowoSIEUx2NQRc8scknde&quot;],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;}},&quot;GmCWdASowoSIEUx2NQRc8scknde&quot;:{&quot;id&quot;:&quot;GmCWdASowoSIEUx2NQRc8scknde&quot;,&quot;snapshot&quot;:{&quot;parent_id&quot;:&quot;UkcMdKC2kouIa8xiSwzccorWnad&quot;,&quot;type&quot;:&quot;text&quot;,&quot;children&quot;:[],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;,&quot;text&quot;:{&quot;initialAttributedTexts&quot;:{&quot;text&quot;:{&quot;0&quot;:&quot;0&quot;},&quot;attribs&quot;:{&quot;0&quot;:&quot;*0+1&quot;}},&quot;apool&quot;:{&quot;numToAttrib&quot;:{&quot;0&quot;:[&quot;author&quot;,&quot;7115230360346050564&quot;]},&quot;nextNum&quot;:1}},&quot;folded&quot;:false}},&quot;FIsKdMaoeo4wSCxYf8fcGzrdnng&quot;:{&quot;id&quot;:&quot;FIsKdMaoeo4wSCxYf8fcGzrdnng&quot;,&quot;snapshot&quot;:{&quot;parent_id&quot;:&quot;HOYad0CMuo2UysxQzSUcvXaVnBb&quot;,&quot;type&quot;:&quot;table_cell&quot;,&quot;children&quot;:[&quot;SyMqdCUyAoAQeuxiWOHcfoarn1g&quot;],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;}},&quot;SyMqdCUyAoAQeuxiWOHcfoarn1g&quot;:{&quot;id&quot;:&quot;SyMqdCUyAoAQeuxiWOHcfoarn1g&quot;,&quot;snapshot&quot;:{&quot;parent_id&quot;:&quot;FIsKdMaoeo4wSCxYf8fcGzrdnng&quot;,&quot;type&quot;:&quot;text&quot;,&quot;children&quot;:[],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;,&quot;text&quot;:{&quot;initialAttributedTexts&quot;:{&quot;text&quot;:{&quot;0&quot;:&quot;16&quot;},&quot;attribs&quot;:{&quot;0&quot;:&quot;*0+2&quot;}},&quot;apool&quot;:{&quot;numToAttrib&quot;:{&quot;0&quot;:[&quot;author&quot;,&quot;7115230360346050564&quot;]},&quot;nextNum&quot;:1}},&quot;folded&quot;:false}},&quot;J0GedGsuoomCC0xsZTyc2h0dnQg&quot;:{&quot;id&quot;:&quot;J0GedGsuoomCC0xsZTyc2h0dnQg&quot;,&quot;snapshot&quot;:{&quot;parent_id&quot;:&quot;HOYad0CMuo2UysxQzSUcvXaVnBb&quot;,&quot;type&quot;:&quot;table_cell&quot;,&quot;children&quot;:[&quot;BCSmdUYwUoyewgxsHIdcwWiBnCd&quot;],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;}},&quot;BCSmdUYwUoyewgxsHIdcwWiBnCd&quot;:{&quot;id&quot;:&quot;BCSmdUYwUoyewgxsHIdcwWiBnCd&quot;,&quot;snapshot&quot;:{&quot;parent_id&quot;:&quot;J0GedGsuoomCC0xsZTyc2h0dnQg&quot;,&quot;type&quot;:&quot;text&quot;,&quot;children&quot;:[],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;,&quot;text&quot;:{&quot;initialAttributedTexts&quot;:{&quot;text&quot;:{&quot;0&quot;:&quot;32&quot;},&quot;attribs&quot;:{&quot;0&quot;:&quot;*0+2&quot;}},&quot;apool&quot;:{&quot;numToAttrib&quot;:{&quot;0&quot;:[&quot;author&quot;,&quot;7115230360346050564&quot;]},&quot;nextNum&quot;:1}},&quot;folded&quot;:false}},&quot;JAW6dMcaCoEqskxKyA5cN5lxncc&quot;:{&quot;id&quot;:&quot;JAW6dMcaCoEqskxKyA5cN5lxncc&quot;,&quot;snapshot&quot;:{&quot;parent_id&quot;:&quot;HOYad0CMuo2UysxQzSUcvXaVnBb&quot;,&quot;type&quot;:&quot;table_cell&quot;,&quot;children&quot;:[&quot;AIYsdwicuoQOyExG40KcLwuGnHd&quot;],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;}},&quot;AIYsdwicuoQOyExG40KcLwuGnHd&quot;:{&quot;id&quot;:&quot;AIYsdwicuoQOyExG40KcLwuGnHd&quot;,&quot;snapshot&quot;:{&quot;parent_id&quot;:&quot;JAW6dMcaCoEqskxKyA5cN5lxncc&quot;,&quot;type&quot;:&quot;text&quot;,&quot;children&quot;:[],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;,&quot;text&quot;:{&quot;initialAttributedTexts&quot;:{&quot;text&quot;:{&quot;0&quot;:&quot;before&quot;},&quot;attribs&quot;:{&quot;0&quot;:&quot;*0+6&quot;}},&quot;apool&quot;:{&quot;numToAttrib&quot;:{&quot;0&quot;:[&quot;author&quot;,&quot;7115230360346050564&quot;]},&quot;nextNum&quot;:1}},&quot;folded&quot;:false}},&quot;S8m0dos80omo4ExLbDtcDLzNnCg&quot;:{&quot;id&quot;:&quot;S8m0dos80omo4ExLbDtcDLzNnCg&quot;,&quot;snapshot&quot;:{&quot;parent_id&quot;:&quot;HOYad0CMuo2UysxQzSUcvXaVnBb&quot;,&quot;type&quot;:&quot;table_cell&quot;,&quot;children&quot;:[&quot;KAW8deQ4woossoxLFWBconyXnhT&quot;],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;}},&quot;KAW8deQ4woossoxLFWBconyXnhT&quot;:{&quot;id&quot;:&quot;KAW8deQ4woossoxLFWBconyXnhT&quot;,&quot;snapshot&quot;:{&quot;type&quot;:&quot;text&quot;,&quot;children&quot;:[],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;,&quot;text&quot;:{&quot;initialAttributedTexts&quot;:{&quot;text&quot;:{&quot;0&quot;:&quot;41.53&quot;},&quot;attribs&quot;:{&quot;0&quot;:&quot;*0+5&quot;}},&quot;apool&quot;:{&quot;numToAttrib&quot;:{&quot;0&quot;:[&quot;author&quot;,&quot;7115230360346050564&quot;]},&quot;nextNum&quot;:1}},&quot;folded&quot;:false,&quot;parent_id&quot;:&quot;S8m0dos80omo4ExLbDtcDLzNnCg&quot;}},&quot;JIIidWUWKoUcKsxWymdcHqwZnre&quot;:{&quot;id&quot;:&quot;JIIidWUWKoUcKsxWymdcHqwZnre&quot;,&quot;snapshot&quot;:{&quot;parent_id&quot;:&quot;HOYad0CMuo2UysxQzSUcvXaVnBb&quot;,&quot;type&quot;:&quot;table_cell&quot;,&quot;children&quot;:[&quot;SYSSdeW0goCSMAxY5Zyc4hUlnPb&quot;],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;}},&quot;SYSSdeW0goCSMAxY5Zyc4hUlnPb&quot;:{&quot;id&quot;:&quot;SYSSdeW0goCSMAxY5Zyc4hUlnPb&quot;,&quot;snapshot&quot;:{&quot;type&quot;:&quot;text&quot;,&quot;children&quot;:[],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;,&quot;text&quot;:{&quot;initialAttributedTexts&quot;:{&quot;text&quot;:{&quot;0&quot;:&quot;57.72&quot;},&quot;attribs&quot;:{&quot;0&quot;:&quot;*0+5&quot;}},&quot;apool&quot;:{&quot;numToAttrib&quot;:{&quot;0&quot;:[&quot;author&quot;,&quot;7115230360346050564&quot;]},&quot;nextNum&quot;:1}},&quot;folded&quot;:false,&quot;parent_id&quot;:&quot;JIIidWUWKoUcKsxWymdcHqwZnre&quot;}},&quot;X8g4dAm2AocGWux8N11cnbz8nsb&quot;:{&quot;id&quot;:&quot;X8g4dAm2AocGWux8N11cnbz8nsb&quot;,&quot;snapshot&quot;:{&quot;parent_id&quot;:&quot;HOYad0CMuo2UysxQzSUcvXaVnBb&quot;,&quot;type&quot;:&quot;table_cell&quot;,&quot;children&quot;:[&quot;T8e6diikkoMS8sxWI5PcarXknme&quot;],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;}},&quot;T8e6diikkoMS8sxWI5PcarXknme&quot;:{&quot;id&quot;:&quot;T8e6diikkoMS8sxWI5PcarXknme&quot;,&quot;snapshot&quot;:{&quot;type&quot;:&quot;text&quot;,&quot;children&quot;:[],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;,&quot;text&quot;:{&quot;initialAttributedTexts&quot;:{&quot;text&quot;:{&quot;0&quot;:&quot;141.95&quot;},&quot;attribs&quot;:{&quot;0&quot;:&quot;*0+6&quot;}},&quot;apool&quot;:{&quot;numToAttrib&quot;:{&quot;0&quot;:[&quot;author&quot;,&quot;7115230360346050564&quot;]},&quot;nextNum&quot;:1}},&quot;folded&quot;:false,&quot;parent_id&quot;:&quot;X8g4dAm2AocGWux8N11cnbz8nsb&quot;}},&quot;DKucduOweo6IGIxIX2ncaOVOn3b&quot;:{&quot;id&quot;:&quot;DKucduOweo6IGIxIX2ncaOVOn3b&quot;,&quot;snapshot&quot;:{&quot;parent_id&quot;:&quot;HOYad0CMuo2UysxQzSUcvXaVnBb&quot;,&quot;type&quot;:&quot;table_cell&quot;,&quot;children&quot;:[&quot;RMKCdwMQUoWC0sxd7xacetygnNH&quot;],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;}},&quot;RMKCdwMQUoWC0sxd7xacetygnNH&quot;:{&quot;id&quot;:&quot;RMKCdwMQUoWC0sxd7xacetygnNH&quot;,&quot;snapshot&quot;:{&quot;parent_id&quot;:&quot;DKucduOweo6IGIxIX2ncaOVOn3b&quot;,&quot;type&quot;:&quot;text&quot;,&quot;children&quot;:[],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;,&quot;text&quot;:{&quot;initialAttributedTexts&quot;:{&quot;text&quot;:{&quot;0&quot;:&quot;after&quot;},&quot;attribs&quot;:{&quot;0&quot;:&quot;*0+5&quot;}},&quot;apool&quot;:{&quot;numToAttrib&quot;:{&quot;0&quot;:[&quot;author&quot;,&quot;7115230360346050564&quot;]},&quot;nextNum&quot;:1}},&quot;folded&quot;:false}},&quot;C4w4diMiioaQ8ExOuGqcZQwLnUc&quot;:{&quot;id&quot;:&quot;C4w4diMiioaQ8ExOuGqcZQwLnUc&quot;,&quot;snapshot&quot;:{&quot;parent_id&quot;:&quot;HOYad0CMuo2UysxQzSUcvXaVnBb&quot;,&quot;type&quot;:&quot;table_cell&quot;,&quot;children&quot;:[&quot;V0QMdAOguoMCMoxiQFJcRPRQnzd&quot;],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;}},&quot;V0QMdAOguoMCMoxiQFJcRPRQnzd&quot;:{&quot;id&quot;:&quot;V0QMdAOguoMCMoxiQFJcRPRQnzd&quot;,&quot;snapshot&quot;:{&quot;type&quot;:&quot;text&quot;,&quot;children&quot;:[],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;,&quot;text&quot;:{&quot;initialAttributedTexts&quot;:{&quot;text&quot;:{&quot;0&quot;:&quot;44.253504 &quot;},&quot;attribs&quot;:{&quot;0&quot;:&quot;*0+a&quot;}},&quot;apool&quot;:{&quot;numToAttrib&quot;:{&quot;0&quot;:[&quot;author&quot;,&quot;7115230360346050564&quot;]},&quot;nextNum&quot;:1}},&quot;folded&quot;:false,&quot;parent_id&quot;:&quot;C4w4diMiioaQ8ExOuGqcZQwLnUc&quot;}},&quot;LSOadeikUo0aSQxa6HTczaWCn1d&quot;:{&quot;id&quot;:&quot;LSOadeikUo0aSQxa6HTczaWCn1d&quot;,&quot;snapshot&quot;:{&quot;parent_id&quot;:&quot;HOYad0CMuo2UysxQzSUcvXaVnBb&quot;,&quot;type&quot;:&quot;table_cell&quot;,&quot;children&quot;:[&quot;SUoSdGE4Co2EYcxKiZAc947Zn9c&quot;],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;}},&quot;SUoSdGE4Co2EYcxKiZAc947Zn9c&quot;:{&quot;id&quot;:&quot;SUoSdGE4Co2EYcxKiZAc947Zn9c&quot;,&quot;snapshot&quot;:{&quot;type&quot;:&quot;text&quot;,&quot;children&quot;:[],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;,&quot;text&quot;:{&quot;initialAttributedTexts&quot;:{&quot;text&quot;:{&quot;0&quot;:&quot;44.71&quot;},&quot;attribs&quot;:{&quot;0&quot;:&quot;*0+5&quot;}},&quot;apool&quot;:{&quot;numToAttrib&quot;:{&quot;0&quot;:[&quot;author&quot;,&quot;7115230360346050564&quot;]},&quot;nextNum&quot;:1}},&quot;folded&quot;:false,&quot;parent_id&quot;:&quot;LSOadeikUo0aSQxa6HTczaWCn1d&quot;}},&quot;J8guduiWWo0soEx6a3zckouenog&quot;:{&quot;id&quot;:&quot;J8guduiWWo0soEx6a3zckouenog&quot;,&quot;snapshot&quot;:{&quot;parent_id&quot;:&quot;HOYad0CMuo2UysxQzSUcvXaVnBb&quot;,&quot;type&quot;:&quot;table_cell&quot;,&quot;children&quot;:[&quot;MSWIdo2csocaeQx0QuNcgrYjnwe&quot;],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;}},&quot;MSWIdo2csocaeQx0QuNcgrYjnwe&quot;:{&quot;id&quot;:&quot;MSWIdo2csocaeQx0QuNcgrYjnwe&quot;,&quot;snapshot&quot;:{&quot;type&quot;:&quot;text&quot;,&quot;children&quot;:[],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;,&quot;text&quot;:{&quot;initialAttributedTexts&quot;:{&quot;text&quot;:{&quot;0&quot;:&quot;51.35&quot;},&quot;attribs&quot;:{&quot;0&quot;:&quot;*0+5&quot;}},&quot;apool&quot;:{&quot;numToAttrib&quot;:{&quot;0&quot;:[&quot;author&quot;,&quot;7115230360346050564&quot;]},&quot;nextNum&quot;:1}},&quot;folded&quot;:false,&quot;parent_id&quot;:&quot;J8guduiWWo0soEx6a3zckouenog&quot;}},&quot;HOYad0CMuo2UysxQzSUcvXaVnBb&quot;:{&quot;id&quot;:&quot;HOYad0CMuo2UysxQzSUcvXaVnBb&quot;,&quot;snapshot&quot;:{&quot;type&quot;:&quot;table&quot;,&quot;children&quot;:[],&quot;comments&quot;:[],&quot;author&quot;:&quot;7115230360346050564&quot;,&quot;columns_id&quot;:[&quot;colf6043245-12f9-4607-be60-9ce3aee2e464&quot;,&quot;col05ac4133-670e-4fc3-83e2-991a1d12dedf&quot;,&quot;colfc79e9d1-8182-46a7-8f40-b60b2d849a40&quot;,&quot;col74fe6183-8186-4555-a292-1a152b271d31&quot;],&quot;rows_id&quot;:[&quot;row37a34112-4976-4f06-aaa1-259dd4c0f385&quot;,&quot;row43959954-737a-42ee-974a-6896a7908c76&quot;,&quot;row8de96261-d61d-41bf-9569-e6df12bc7182&quot;],&quot;column_set&quot;:{&quot;colf6043245-12f9-4607-be60-9ce3aee2e464&quot;:{&quot;column_width&quot;:100},&quot;col05ac4133-670e-4fc3-83e2-991a1d12dedf&quot;:{&quot;column_width&quot;:100},&quot;colfc79e9d1-8182-46a7-8f40-b60b2d849a40&quot;:{&quot;column_width&quot;:100},&quot;col74fe6183-8186-4555-a292-1a152b271d31&quot;:{&quot;column_width&quot;:100}},&quot;cell_set&quot;:{&quot;row37a34112-4976-4f06-aaa1-259dd4c0f385colf6043245-12f9-4607-be60-9ce3aee2e464&quot;:{&quot;block_id&quot;:&quot;NgO0dCYuSo0SMQxSO9TcDO6gn0e&quot;,&quot;merge_info&quot;:{&quot;row_span&quot;:1,&quot;col_span&quot;:1}},&quot;row37a34112-4976-4f06-aaa1-259dd4c0f385col05ac4133-670e-4fc3-83e2-991a1d12dedf&quot;:{&quot;block_id&quot;:&quot;UkcMdKC2kouIa8xiSwzccorWnad&quot;,&quot;merge_info&quot;:{&quot;row_span&quot;:1,&quot;col_span&quot;:1}},&quot;row37a34112-4976-4f06-aaa1-259dd4c0f385colfc79e9d1-8182-46a7-8f40-b60b2d849a40&quot;:{&quot;block_id&quot;:&quot;FIsKdMaoeo4wSCxYf8fcGzrdnng&quot;,&quot;merge_info&quot;:{&quot;row_span&quot;:1,&quot;col_span&quot;:1}},&quot;row37a34112-4976-4f06-aaa1-259dd4c0f385col74fe6183-8186-4555-a292-1a152b271d31&quot;:{&quot;block_id&quot;:&quot;J0GedGsuoomCC0xsZTyc2h0dnQg&quot;,&quot;merge_info&quot;:{&quot;row_span&quot;:1,&quot;col_span&quot;:1}},&quot;row43959954-737a-42ee-974a-6896a7908c76colf6043245-12f9-4607-be60-9ce3aee2e464&quot;:{&quot;block_id&quot;:&quot;JAW6dMcaCoEqskxKyA5cN5lxncc&quot;,&quot;merge_info&quot;:{&quot;row_span&quot;:1,&quot;col_span&quot;:1}},&quot;row43959954-737a-42ee-974a-6896a7908c76col05ac4133-670e-4fc3-83e2-991a1d12dedf&quot;:{&quot;block_id&quot;:&quot;S8m0dos80omo4ExLbDtcDLzNnCg&quot;,&quot;merge_info&quot;:{&quot;row_span&quot;:1,&quot;col_span&quot;:1}},&quot;row43959954-737a-42ee-974a-6896a7908c76colfc79e9d1-8182-46a7-8f40-b60b2d849a40&quot;:{&quot;block_id&quot;:&quot;JIIidWUWKoUcKsxWymdcHqwZnre&quot;,&quot;merge_info&quot;:{&quot;row_span&quot;:1,&quot;col_span&quot;:1}},&quot;row43959954-737a-42ee-974a-6896a7908c76col74fe6183-8186-4555-a292-1a152b271d31&quot;:{&quot;block_id&quot;:&quot;X8g4dAm2AocGWux8N11cnbz8nsb&quot;,&quot;merge_info&quot;:{&quot;row_span&quot;:1,&quot;col_span&quot;:1}},&quot;row8de96261-d61d-41bf-9569-e6df12bc7182colf6043245-12f9-4607-be60-9ce3aee2e464&quot;:{&quot;block_id&quot;:&quot;DKucduOweo6IGIxIX2ncaOVOn3b&quot;,&quot;merge_info&quot;:{&quot;row_span&quot;:1,&quot;col_span&quot;:1}},&quot;row8de96261-d61d-41bf-9569-e6df12bc7182col05ac4133-670e-4fc3-83e2-991a1d12dedf&quot;:{&quot;block_id&quot;:&quot;C4w4diMiioaQ8ExOuGqcZQwLnUc&quot;,&quot;merge_info&quot;:{&quot;row_span&quot;:1,&quot;col_span&quot;:1}},&quot;row8de96261-d61d-41bf-9569-e6df12bc7182colfc79e9d1-8182-46a7-8f40-b60b2d849a40&quot;:{&quot;block_id&quot;:&quot;LSOadeikUo0aSQxa6HTczaWCn1d&quot;,&quot;merge_info&quot;:{&quot;row_span&quot;:1,&quot;col_span&quot;:1}},&quot;row8de96261-d61d-41bf-9569-e6df12bc7182col74fe6183-8186-4555-a292-1a152b271d31&quot;:{&quot;block_id&quot;:&quot;J8guduiWWo0soEx6a3zckouenog&quot;,&quot;merge_info&quot;:{&quot;row_span&quot;:1,&quot;col_span&quot;:1}}},&quot;parent_id&quot;:&quot;CDC9dn54Voz89KxfQGScj6zanMb&quot;}}},&quot;payloadMap&quot;:{&quot;AWyIdU0aYossGwxbHVXcNszSnbN&quot;:{&quot;level&quot;:1},&quot;GmCWdASowoSIEUx2NQRc8scknde&quot;:{&quot;level&quot;:1},&quot;SyMqdCUyAoAQeuxiWOHcfoarn1g&quot;:{&quot;level&quot;:1},&quot;BCSmdUYwUoyewgxsHIdcwWiBnCd&quot;:{&quot;level&quot;:1},&quot;AIYsdwicuoQOyExG40KcLwuGnHd&quot;:{&quot;level&quot;:1},&quot;KAW8deQ4woossoxLFWBconyXnhT&quot;:{&quot;level&quot;:1},&quot;SYSSdeW0goCSMAxY5Zyc4hUlnPb&quot;:{&quot;level&quot;:1},&quot;T8e6diikkoMS8sxWI5PcarXknme&quot;:{&quot;level&quot;:1},&quot;RMKCdwMQUoWC0sxd7xacetygnNH&quot;:{&quot;level&quot;:1},&quot;V0QMdAOguoMCMoxiQFJcRPRQnzd&quot;:{&quot;level&quot;:1},&quot;SUoSdGE4Co2EYcxKiZAc947Zn9c&quot;:{&quot;level&quot;:1},&quot;MSWIdo2csocaeQx0QuNcgrYjnwe&quot;:{&quot;level&quot;:1}},&quot;extra&quot;:{&quot;mention_page_title&quot;:{},&quot;external_mention_url&quot;:{}},&quot;isKeepQuoteContainer&quot;:false,&quot;selection&quot;:[{&quot;id&quot;:640,&quot;type&quot;:&quot;block&quot;,&quot;recordId&quot;:&quot;NgO0dCYuSo0SMQxSO9TcDO6gn0e&quot;},{&quot;id&quot;:642,&quot;type&quot;:&quot;block&quot;,&quot;recordId&quot;:&quot;UkcMdKC2kouIa8xiSwzccorWnad&quot;},{&quot;id&quot;:644,&quot;type&quot;:&quot;block&quot;,&quot;recordId&quot;:&quot;FIsKdMaoeo4wSCxYf8fcGzrdnng&quot;},{&quot;id&quot;:646,&quot;type&quot;:&quot;block&quot;,&quot;recordId&quot;:&quot;J0GedGsuoomCC0xsZTyc2h0dnQg&quot;},{&quot;id&quot;:648,&quot;type&quot;:&quot;block&quot;,&quot;recordId&quot;:&quot;JAW6dMcaCoEqskxKyA5cN5lxncc&quot;},{&quot;id&quot;:650,&quot;type&quot;:&quot;block&quot;,&quot;recordId&quot;:&quot;S8m0dos80omo4ExLbDtcDLzNnCg&quot;},{&quot;id&quot;:652,&quot;type&quot;:&quot;block&quot;,&quot;recordId&quot;:&quot;JIIidWUWKoUcKsxWymdcHqwZnre&quot;},{&quot;id&quot;:654,&quot;type&quot;:&quot;block&quot;,&quot;recordId&quot;:&quot;X8g4dAm2AocGWux8N11cnbz8nsb&quot;},{&quot;id&quot;:656,&quot;type&quot;:&quot;block&quot;,&quot;recordId&quot;:&quot;DKucduOweo6IGIxIX2ncaOVOn3b&quot;},{&quot;id&quot;:658,&quot;type&quot;:&quot;block&quot;,&quot;recordId&quot;:&quot;C4w4diMiioaQ8ExOuGqcZQwLnUc&quot;},{&quot;id&quot;:660,&quot;type&quot;:&quot;block&quot;,&quot;recordId&quot;:&quot;LSOadeikUo0aSQxa6HTczaWCn1d&quot;},{&quot;id&quot;:662,&quot;type&quot;:&quot;block&quot;,&quot;recordId&quot;:&quot;J8guduiWWo0soEx6a3zckouenog&quot;}],&quot;pasteFlag&quot;:&quot;d7a781e6-55d1-42e5-b338-cf4ab4937f88&quot;}" data-lark-record-format="docx/record" class="lark-record-clipboard"></span>

## Checklist:

- [x] I have added test cases for my bug fix or my new feature (munually test by setting `local_exchange_buffer_mem_limit_per_driver` to 1 to mock woking on large chunks)
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Review by murphyatwork — DISMISSED** [review]

**Review by silverbullet233 — DISMISSED** [review]
LGTM

**Review by kangkaisen — DISMISSED** [review]

**Review by silverbullet233 — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

File: be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp [hunk 1]
```diff
@@ -113,12 +115,17 @@ ChunkPtr LocalExchangeSourceOperator::_pull_shuffle_chunk(RuntimeState* state) {
         while (!_partition_chunk_queue.empty() &&
                rows_num + _partition_chunk_queue.front().size <= state->chunk_size()) {
             rows_num += _partition_chunk_queue.front().size;
+            memory_size += _partition_chunk_queue.front().chunk->bytes_usage(_partition_chunk_queue.front().from,
+                                                                             _partition_chunk_queue.front().size);
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp:119]: This shuffled chunk doesn't uses `[from, size)`, but the indexes `partition_chunk.indexes->data()[from]`, `partition_chunk.indexes->data()[from+1]`, ...`partition_chunk.indexes->data()[from + size -1 ]`.

**Review by satanson — DISMISSED** [review]

File: be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp [hunk 2]
```diff
@@ -75,7 +76,7 @@ Status LocalExchangeSourceOperator::set_finished(RuntimeState* state) {
     // clear _partition_chunk_queue
     { [[maybe_unused]] typeof(_partition_chunk_queue) tmp = std::move(_partition_chunk_queue); }
     // Subtract the number of rows of buffered chunks from row_count of _memory_manager and make it unblocked.
-    _memory_manager->update_row_count(-(full_rows_num + _partition_rows_num));
+    _memory_manager->update_memory_usage(-_memory_manager->get_memory_usage());
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp:79]: Should only decrease the memory usage of this `LocalExchangeSourceOperator`, but `_memory_manager->get_memory_usage()` is the memory of all the `LocalExchangeSourceOperator`s.

File: be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp (hunk 1 above)
Comment (fzhedu) [reply be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp]: my mistake, fixed by actually computing partition memory usage.

File: be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp (hunk 2 above)
Comment (fzhedu) [reply be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp]: my mistake, fixed by introducing  `_local_memory_usage`.

**Review by ZiheLiu — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Pull request merged**